crates/egui_kittest/src/snapshot.rs

```diff
@@ -0,0 +1,122 @@
+use image::ImageError;
+use std::fmt::Display;
+use std::io::ErrorKind;
+use std::path::{Path, PathBuf};
+
+#[derive(Debug)]
+pub enum SnapshotError {
+    Diff {
+        diff: i32,
+        diff_path: PathBuf,
+    },
+    OpenSnapshot {
+        path: PathBuf,
+        err: image::ImageError,
+    },
+    SizeMismatch {
+        expected: (u32, u32),
+        actual: (u32, u32),
+    },
+}
+
+impl Display for SnapshotError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Diff { diff, diff_path } => {
+                write!(
+                    f,
+                    "Image did not match snapshot. Diff: {diff}, {diff_path:?}"
+                )
+            }
+            Self::OpenSnapshot { path, err } => match err {
+                ImageError::IoError(io) => match io.kind() {
+                    ErrorKind::NotFound => {
+                        write!(f, "Missing snapshot: {path:?}")
+                    }
+                    err => {
+                        write!(f, "Error reading snapshot: {err:?}")
+                    }
+                },
+                err => {
+                    write!(f, "Error decoding snapshot: {err:?}")
+                }
+            },
+            Self::SizeMismatch { expected, actual } => {
+                write!(
+                    f,
+                    "Image size did not match snapshot. Expected: {expected:?}, Actual: {actual:?}"
+                )
+            }
+        }
+    }
+}
+
+/// Image snapshot test.
+///
+/// # Errors
+/// Returns a [`SnapshotError`] if the image does not match the snapshot.
+pub fn try_image_snapshot(current: &image::RgbaImage, name: &str) -> Result<(), SnapshotError> {
+    let snapshots_path = Path::new("tests/snapshots");
+
+    let path = snapshots_path.join(format!("{name}.png"));
+    std::fs::create_dir_all(path.parent().unwrap()).ok();
+
+    let diff_path = snapshots_path.join(format!("{name}.diff.png"));
+    let current_path = snapshots_path.join(format!("{name}.new.png"));
+
+    current.save(&current_path).unwrap();
```

Comment:
Maybe failure to write images to disk should also result in an error, rather than a panic.

In either case, it should include the problematic path in the error message